outlookmail/streams.bal

```diff
@@ -0,0 +1,121 @@
+// Copyright (c) 2021 WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+//
+// WSO2 Inc. licenses this file to you under the Apache License,
+// Version 2.0 (the "License"); you may not use this file except
+// in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+import ballerina/http;
+
+class MessageStream {
+    private Message[] messageEntries = [];
+    int index = 0;
+    string nextLink = EMPTY_STRING;
+    Configuration config;
+
+    public isolated function init(json payload, Configuration config) returns @tainted error? {
```

Comment:
There was a problem in streams when we have `$top` parameter. This is because although we have `$top` defined, the `nextLink` is returned in the response. As we only check the necessity of a `nextLink` to continue the data pulling in the stream although a user include a $top parameter in the request all the records are returned. This issue was not there in MS Excel. But was in MS Teams and Outlook calendar. Check if same applies for this too. If so sample code of what we have used can be found [here](https://github.com/sachinira/module-ballerinax-azure.ad/blob/beta1/aad/stream_implementer.bal#L45)

Reply:
**I think, $top is used to limit the pagesize, but not to restrict the number of top records we need to get in real use.** So what we do is that we change the number of records that come per one page, then the number of requests will vary. Whatever the top value is, we should get the whole results filtered through other parameters. [Please refer here to query parameters](https://docs.microsoft.com/en-us/graph/query-parameters#top-parameter).  

> Depending on the page size and mailbox data, getting messages from a mailbox can incur multiple requests. The default page size is 10 messages. Use $top to customize the page size, within the range of 1 and 1000.

**According to the above what if a user needs the top 1500 records. So User should get the top 1500, but the page size(top value) should be lesser than 1000.**

Therefore, what he can do is that putting a limit to the stream when he uses it. User might use,

1. loop breaks in stream foreach
2. consume the stream using the [next](https://docs.central.ballerina.io/ballerina/lang.stream/0.4.0/functions#next) function of the stream module. then he can stop wherever the total count he wants.
Due to the above reason, I **think, a user will not be able to get more than 1000 records if we use the suggestion you have mentioned [here]**(https://github.com/sachinira/module-ballerinax-azure.ad/blob/beta1/aad/stream_implementer.bal#L45). and also if we don't tune up the query params, then there might be gateway timeout too as the following explanation 

> To improve the operation response time, use $select to specify the exact properties you need;  Fine-tune the values for $select and $top, especially when you must use a larger page size, as returning a page with hundreds of messages each with a full response payload may trigger the gateway timeout (HTTP 504).

Reply:
There were several things  in my mind to have the code like this

**[1]** There were some good use cases for `$top` as I think. 
Think of a scenario that user wants to get the 5 records skipping the first 10 records of the original result set (He can use `$top` in combination with `$skip`) We can't do this from the stream's side because we don't have the option to skip in the stream. In that case using `$top` may come in handy. 
We can of course use counting method here too. But the code written by user will be less if we limit the top.

**[2]** In the scenario if user specifies more than the limit (ex: page size limit is 1 to 1000 and he specifies $top=1500) for the top he will get an error for invalid page size anyways and it will not return any results. So for that case using the stream to limit the results is good

 **[4]** The expected usage of `$top` in Graph API will be removed because we have to keep a count in user code and do the things.

**[3]** About fine tuning the response, it is out of scope of the problem here.

Reply:
Yeah, IMO, 
[1] Microsoft does not recommend manipulating data using `$skip`. 

> Do not try to extract the $skip value from the @odata.nextLink URL to manipulate responses. This API uses the $skip value to keep count of all the items it has gone through in the user's mailbox to return a page of message-type items. It's therefore possible that even in the initial response, the $skip value is larger than the page size. For more information

[2] Yes, If someone gives other values than they support. It is clear that they won't send any results. it is not the real use of the $top, getting the first set of records. but setting the max record count per request.
[3] I think, removing the Microsoft expected usage of $top is not a better way to use their PR.
[4] $top and $select we use to tune up the response, because if someone has a large number of mails in a mailbox. if he wants to get them, he has to add tune-up to the query otherwise operation won't be successful. so we need to support it in our connectors. not to send the extract $top number of records but to send all records he wants with the page size he wants. so he can also make the decision on how many he wants, but not by us.

Reply:
I'll add your suggestion and thank you very much for your support :+1: